spmd/tensor/placement_types.py

```diff
@@ -30,6 +32,170 @@ class Shard(Placement):
+                if (
+                    with_padding
+                    and idx_start_to_pad != 0
+                    and i >= idx_start_to_pad
+                ):
+                    shard = self._pad_tensor(shard)
+                # input tensors are expected to be congtiguous by the collective backend
+                shard = shard.contiguous() if contiguous else shard
+                shard_list.append(shard)
+            return shard_list, idx_start_to_pad
+        else:
+            return tensor_list, idx_start_to_pad
+
+    def _pad_tensor(self, tensor: torch.Tensor) -> torch.Tensor:
+        # pad tensor by 1 on the shard dim
+        pad = [0, 0] * (tensor.ndim - self.dim)
+        pad[-1] = 1
+        return torch.nn.functional.pad(tensor, pad)
+
+    def _unpad_tensor(self, tensor: torch.Tensor) -> torch.Tensor:
+        # unpad tensor by 1 on the shard dim
+        return tensor.narrow(
+            self.dim, start=0, length=tensor.size(self.dim) - 1
+        )
+
+    def local_shard_size_on_dim(
+        self,
+        size_on_dim: int,
+        num_chunks: int,
+        rank: int,
+        return_offset: bool = False,
+    ) -> Tuple[int, int]:
+        # return the local shard size and offset on a given tensor dim
+        assert (
+            size_on_dim >= num_chunks
+        ), f"Size to be sharded on dim {self.dim} must be at least as large as the number of devices in that dimension {num_chunks}"
+        split_size, pad_idx = divmod(size_on_dim, num_chunks)
+        local_shard_size = (
+            split_size + 1 if pad_idx != 0 and rank < pad_idx else split_size
+        )
+        local_offset_on_dim = -1
+        if return_offset:
+            local_offset_on_dim = (
+                rank * split_size + pad_idx if rank >= pad_idx else rank
+            )
+        return (local_shard_size, local_offset_on_dim)
+
+    def shard_tensor(
+        self, tensor: torch.Tensor, mesh: DeviceMesh, mesh_dim: int
+    ) -> torch.Tensor:
+        # shard and scatter a tensor on a mesh dimension (use coordinate
+        # 0 on the mesh dimension as source of truth)
+        my_coordinate = mesh.get_coordinate_on_dim(mesh_dim)
+        num_chunks = mesh.size(dim=mesh_dim)
+        # TODO: what should happen if rank is not in the mesh?
+        # see issue https://github.com/pytorch/tau/pull/492
+        assert (
+            my_coordinate is not None
+        ), "Rank if not part of mesh"  # TODO: figure out behavior here
+        scatter_list, pad_idx = self._split_tensor(
+            tensor, num_chunks, with_padding=True, contiguous=True
+        )
+        output = torch.empty_like(scatter_list[my_coordinate])
+        mesh.scatter(output, scatter_list, mesh_dim=mesh_dim)
+
+        if pad_idx != 0 and my_coordinate >= pad_idx:
+            output = self._unpad_tensor(output)
+        return output
+
+    def reduce_shard_tensor(
+        self, tensor: torch.Tensor, mesh: DeviceMesh, mesh_dim: int
```

Comment:
It feels a bit weird that this is a member of Shard class and that it takes a _non_ DTensor. 
1) does this need to be a member of Shard?
2) what really are we expecting/allowing as "tensor" argument? Is it supposed to be one _shard_ of a DTensor? i can see that but this would at most be ok as a private API. Seems too confusing/risky for end-user to use.

Reply:
The reason I keep all the sharding related logic into `Shard` directly is that the way to do sharding `tensor_split`, and padding with `tensor_split` are specific to `tensor_split`, if we want to do another type of sharding, it's easy to just extend the `Placement` directly and implement some sharding logic that is required by `distribute_tensor` and `redistribute`, (i.e. for interleaved sharding we can easily implement a interleaved placement that would work). 

Agree that this is actually all internal logic and we should not expose this to end-user, I'll make the methods be all private.